File: C#/GigaChadSys/GigaChadSys-DAO/EntrenadorDAOImpl.cs
using GigaChadSysDBManager;
using GigaChadSysModel;
using MySql.Data.MySqlClient;

namespace GigaChadSysDAO;

public class EntrenadorDAOImpl : EntrenadorDAO
{
    public List<Entrenador> ListAll()
    {
        List<Entrenador> entrenadores = new();

        string sql = @"SELECT idUsuario, nombreCompleto, dni, email, telefono,
                              contrasenia, especialidad, sueldo, tiempoTrabajo, activo
                       FROM Entrenador
                       WHERE activo = 1";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);
            using MySqlDataReader reader = comando.ExecuteReader();

            while (reader.Read())
            {
                entrenadores.Add(MapearEntrenador(reader));
            }
        }
        catch (Exception ex)
        {
            throw new Exception("Error al listar entrenadores.", ex);
        }

        return entrenadores;
    }

    public Entrenador? Load(int id)
    {
        Entrenador? entrenador = null;

        string sql = @"SELECT idUsuario, nombreCompleto, dni, email, telefono,
                              contrasenia, especialidad, sueldo, tiempoTrabajo, activo
                       FROM Entrenador
                       WHERE idUsuario = @idUsuario";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);

            comando.Parameters.AddWithValue("@idUsuario", id);

            using MySqlDataReader reader = comando.ExecuteReader();

            if (reader.Read())
            {
                entrenador = MapearEntrenador(reader);
            }
        }
        catch (Exception ex)
        {
            throw new Exception("Error al cargar entrenador.", ex);
        }

        return entrenador;
    }

    public Entrenador Save(Entrenador entrenador)
    {
        entrenador.Activo = true;

        string sql = @"INSERT INTO Entrenador
                       (nombreCompleto, dni, email, telefono, contrasenia,
                        especialidad, sueldo, tiempoTrabajo, activo)
                       VALUES
                       (@nombreCompleto, @dni, @email, @telefono, @contrasenia,
                        @especialidad, @sueldo, @tiempoTrabajo, @activo)";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);

            comando.Parameters.AddWithValue("@nombreCompleto", entrenador.NombreCompleto);
            comando.Parameters.AddWithValue("@dni", entrenador.Dni);
            comando.Parameters.AddWithValue("@email", entrenador.Email);
            comando.Parameters.AddWithValue("@telefono", entrenador.Telefono);
            comando.Parameters.AddWithValue("@contrasenia", entrenador.Contrasenia);
            comando.Parameters.AddWithValue("@especialidad", entrenador.Especialidad);
            comando.Parameters.AddWithValue("@sueldo", entrenador.Sueldo);
            comando.Parameters.AddWithValue("@tiempoTrabajo", entrenador.TiempoTrabajo);
            comando.Parameters.AddWithValue("@activo", entrenador.Activo);

            comando.ExecuteNonQuery();

            entrenador.IdUsuario = (int)comando.LastInsertedId;
        }
        catch (Exception ex)
        {
            throw new Exception("Error al registrar entrenador.", ex);
        }

        return entrenador;
    }

    public Entrenador? Update(Entrenador entrenador)
    {
        string sql = @"UPDATE Entrenador SET
                            nombreCompleto = @nombreCompleto,
                            dni = @dni,
                            email = @email,
                            telefono = @telefono,
                            contrasenia = @contrasenia,
                            especialidad = @especialidad,
                            sueldo = @sueldo,
                            tiempoTrabajo = @tiempoTrabajo,
                            activo = @activo
                       WHERE idUsuario = @idUsuario";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);

            comando.Parameters.AddWithValue("@nombreCompleto", entrenador.NombreCompleto);
            comando.Parameters.AddWithValue("@dni", entrenador.Dni);
            comando.Parameters.AddWithValue("@email", entrenador.Email);
            comando.Parameters.AddWithValue("@telefono", entrenador.Telefono);
            comando.Parameters.AddWithValue("@contrasenia", entrenador.Contrasenia);
            comando.Parameters.AddWithValue("@especialidad", entrenador.Especialidad);
            comando.Parameters.AddWithValue("@sueldo", entrenador.Sueldo);
            comando.Parameters.AddWithValue("@tiempoTrabajo", entrenador.TiempoTrabajo);
            comando.Parameters.AddWithValue("@activo", entrenador.Activo);
            comando.Parameters.AddWithValue("@idUsuario", entrenador.IdUsuario);

            int filas = comando.ExecuteNonQuery();

            return filas > 0 ? entrenador : null;
        }
        catch (Exception ex)
        {
            throw new Exception("Error al actualizar entrenador.", ex);
        }
    }

    public void Remove(Entrenador entrenador)
    {
        entrenador.Activo = false;

        string sql = @"UPDATE Entrenador
                       SET activo = @activo
                       WHERE idUsuario = @idUsuario";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);

            comando.Parameters.AddWithValue("@activo", entrenador.Activo);
            comando.Parameters.AddWithValue("@idUsuario", entrenador.IdUsuario);

            comando.ExecuteNonQuery();
        }
        catch (Exception ex)
        {
            throw new Exception("Error al eliminar entrenador.", ex);
        }
    }

    private Entrenador MapearEntrenador(MySqlDataReader reader)
    {
        return new Entrenador
        {
            IdUsuario = reader.GetInt32("idUsuario"),
            NombreCompleto = reader.GetString("nombreCompleto"),
            Dni = reader.IsDBNull(reader.GetOrdinal("dni")) ? "" : reader.GetString("dni"),
            Email = reader.GetString("email"),
            Telefono = reader.IsDBNull(reader.GetOrdinal("telefono")) ? "" : reader.GetString("telefono"),
            Contrasenia = reader.GetString("contrasenia"),
            Especialidad = reader.IsDBNull(reader.GetOrdinal("especialidad")) ? "" : reader.GetString("especialidad"),
            Sueldo = reader.GetDecimal("sueldo"),
            TiempoTrabajo = reader.GetTimeSpan("tiempoTrabajo"),
            Activo = reader.GetBoolean("activo")
        };
    }

    public Entrenador? LoadByEmail(string email)
    {
        Entrenador? entrenador = null;

        string sql = @"SELECT idUsuario, nombreCompleto, dni, email, telefono,
                          contrasenia, especialidad, sueldo, tiempoTrabajo, activo
                   FROM Entrenador
                   WHERE email = @email AND activo = 1";

        try
        {
            using MySqlConnection conexion = DBManager.Instance.GetConnection();
            using MySqlCommand comando = new(sql, conexion);

            comando.Parameters.AddWithValue("@email", email);

            using MySqlDataReader reader = comando.ExecuteReader();

            if (reader.Read())
            {
                entrenador = MapearEntrenador(reader);
            }
        }
        catch (Exception ex)
        {
            throw new Exception("Error al buscar entrenador por email.", ex);
        }

        return entrenador;
    }
}

File: Java/GigaChadSys-DAO/src/main/java/pe/edu/pucp/gigachadsys/dao/impl/usuarios/EntrenadorDAOImpl.java
package pe.edu.pucp.gigachadsys.dao.impl.usuarios;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

import pe.edu.pucp.gigachadsys.dao.inter.usuarios.EntrenadorDAO;
import pe.edu.pucp.gigachadsys.model.personas.Entrenador;
import pe.edu.pucp.gigachadsys.dao.manager.DBManager;

public class EntrenadorDAOImpl implements EntrenadorDAO {

    @Override
    public List<Entrenador> listAll() {
        List<Entrenador> list = new ArrayList<>();

        String sql =
                "SELECT " +
                        "idUsuario, " +
                        "nombres, " +
                        "apellidoPaterno, " +
                        "apellidoMaterno, " +
                        "edad, " +
                        "DNI, " +
                        "email, " +
                        "telefono, " +
                        "contrasenia, " +
                        "rol, " +
                        "Especialidad, " +
                        "sueldo, " +
                        "tiempoTrabajo, " +
                        "activo " +
                        "FROM Entrenador";

        try (
                Connection connection = DBManager.getInstance().getConnection();
                Statement stm = connection.createStatement();
                ResultSet rs = stm.executeQuery(sql)
        ) {
            while (rs.next()) {
                list.add(mapearEntrenador(rs));
            }

            return list;

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Entrenador load(Integer id) {
        String sql =
                "SELECT " +
                        "idUsuario, " +
                        "nombres, " +
                        "apellidoPaterno, " +
                        "apellidoMaterno, " +
                        "edad, " +
                        "DNI, " +
                        "email, " +
                        "telefono, " +
                        "contrasenia, " +
                        "rol, " +
                        "Especialidad, " +
                        "sueldo, " +
                        "tiempoTrabajo, " +
                        "activo " +
                        "FROM Entrenador " +
                        "WHERE idUsuario = ?";

        try (
                Connection connection = DBManager.getInstance().getConnection();
                PreparedStatement pstmt = connection.prepareStatement(sql)
        ) {
            pstmt.setInt(1, id);

            try (ResultSet rs = pstmt.executeQuery()) {
                if (rs.next()) {
                    return mapearEntrenador(rs);
                }
            }

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return null;
    }

    @Override
    public Entrenador save(Entrenador entrenador) {
        entrenador.setActivo(true);

        String sql =
                "INSERT INTO Entrenador (" +
                        "nombres, " +
                        "apellidoPaterno, " +
                        "apellidoMaterno, " +
                        "edad, " +
                        "DNI, " +
                        "email, " +
                        "telefono, " +
                        "contrasenia, " +
                        "rol, " +
                        "Especialidad, " +
                        "sueldo, " +
                        "tiempoTrabajo, " +
                        "activo" +
                        ") VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";

        try (
                Connection connection = DBManager.getInstance().getConnection();
                PreparedStatement pstmt = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)
        ) {
            pstmt.setString(1, entrenador.getNombres());
            pstmt.setString(2, entrenador.getApellidoPaterno());
            pstmt.setString(3, entrenador.getApellidoMaterno());
            pstmt.setInt(4, entrenador.getEdad());
            pstmt.setInt(5, entrenador.getDni());
            pstmt.setString(6, entrenador.getEmail());
            pstmt.setInt(7, entrenador.getTelefono());
            pstmt.setString(8, entrenador.getContrasenia());
            pstmt.setString(9, entrenador.getRol() == null || entrenador.getRol().isBlank() ? "Entrenador" : entrenador.getRol());
            pstmt.setString(10, entrenador.getEspecialidad());
            pstmt.setDouble(11, entrenador.getSueldo());
            pstmt.setTime(12, entrenador.getTiempoTrabajado());
            pstmt.setBoolean(13, entrenador.getActivo() != null ? entrenador.getActivo() : true);

            int affectedRows = pstmt.executeUpdate();

            if (affectedRows > 0) {
                try (ResultSet generatedKeys = pstmt.getGeneratedKeys()) {
                    if (generatedKeys.next()) {
                        entrenador.setIdUsuario(generatedKeys.getInt(1));
                    }
                }
            }

            return entrenador;

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Entrenador update(Entrenador entrenador) {
        String sql =
                "UPDATE Entrenador SET " +
                        "nombres = ?, " +
                        "apellidoPaterno = ?, " +
                        "apellidoMaterno = ?, " +
                        "edad = ?, " +
                        "DNI = ?, " +
                        "email = ?, " +
                        "telefono = ?, " +
                        "contrasenia = COALESCE(NULLIF(?, ''), contrasenia), " +
                        "rol = ?, " +
                        "Especialidad = ?, " +
                        "sueldo = ?, " +
                        "tiempoTrabajo = ?, " +
                        "activo = ? " +
                        "WHERE idUsuario = ?";

        try (
                Connection connection = DBManager.getInstance().getConnection();
                PreparedStatement pstmt = connection.prepareStatement(sql)
        ) {
            pstmt.setString(1, entrenador.getNombres());
            pstmt.setString(2, entrenador.getApellidoPaterno());
            pstmt.setString(3, entrenador.getApellidoMaterno());
            pstmt.setInt(4, entrenador.getEdad());
            pstmt.setInt(5, entrenador.getDni());
            pstmt.setString(6, entrenador.getEmail());
            pstmt.setInt(7, entrenador.getTelefono());
            pstmt.setString(8, entrenador.getContrasenia());
            pstmt.setString(9, entrenador.getRol() == null || entrenador.getRol().isBlank() ? "Entrenador" : entrenador.getRol());
            pstmt.setString(10, entrenador.getEspecialidad());
            pstmt.setDouble(11, entrenador.getSueldo());
            pstmt.setTime(12, entrenador.getTiempoTrabajado());
            pstmt.setBoolean(13, entrenador.getActivo() != null ? entrenador.getActivo() : true);
            pstmt.setInt(14, entrenador.getIdUsuario());

            int affectedRows = pstmt.executeUpdate();

            if (affectedRows > 0) {
                return entrenador;
            }

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return null;
    }

    @Override
    public void remove(Entrenador entrenador) {
        entrenador.setActivo(false);

        String sql =
                "UPDATE Entrenador " +
                        "SET activo = ? " +
                        "WHERE idUsuario = ?";

        try (
                Connection connection = DBManager.getInstance().getConnection();
                PreparedStatement pstmt = connection.prepareStatement(sql)
        ) {
            pstmt.setBoolean(1, entrenador.getActivo());
            pstmt.setInt(2, entrenador.getIdUsuario());

            pstmt.executeUpdate();

        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private Entrenador mapearEntrenador(ResultSet rs) throws SQLException {
        Entrenador entrenador = new Entrenador();

        entrenador.setIdUsuario(rs.getInt("idUsuario"));
        entrenador.setNombres(rs.getString("nombres"));
        entrenador.setApellidoPaterno(rs.getString("apellidoPaterno"));
        entrenador.setApellidoMaterno(rs.getString("apellidoMaterno"));
        entrenador.setEdad(rs.getInt("edad"));
        entrenador.setDni(rs.getInt("DNI"));
        entrenador.setEmail(rs.getString("email"));
        entrenador.setTelefono(rs.getInt("telefono"));
        entrenador.setContrasenia(rs.getString("contrasenia"));
        entrenador.setRol(rs.getString("rol"));
        entrenador.setEspecialidad(rs.getString("Especialidad"));
        entrenador.setSueldo(rs.getDouble("sueldo"));
        entrenador.setTiempoTrabajado(rs.getTime("tiempoTrabajo"));
        entrenador.setActivo(rs.getBoolean("activo"));

        return entrenador;
    }
}